services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/VersionedRecordExtension.java

```diff
@@ -142,13 +159,131 @@ public WriteModification beforeWrite(DynamoDbExtensionContext.BeforeWrite contex
+                                         // Check that the version does not exist before setting the initial value.
+                                         .expression(String.format("attribute_not_exists(%s)", attributeKeyRef))
+                                         .expressionNames(Collections.singletonMap(attributeKeyRef, versionAttributeKey))
+                                         .build();
+
+        return Pair.of(newVersionValue, condition);
+    }
+
+    private Pair<AttributeValue, Expression> updateExistingRecord(String versionAttributeKey,
+                                                                  AttributeValue existingVersionValue,
+                                                                  DynamoDbExtensionContext.BeforeWrite context) {
+        int existingVersion = getExistingVersion(existingVersionValue);
+        AttributeValue newVersionValue = incrementVersion(existingVersion, context);
+
+        String attributeKeyRef = keyRef(versionAttributeKey);
+        String existingVersionValueKey = VERSIONED_RECORD_EXPRESSION_VALUE_KEY_MAPPER.apply(versionAttributeKey);
+
+        Expression condition = Expression.builder()
+                                         // Check that the version matches the existing value before setting the updated value.
+                                         .expression(String.format("%s = %s", attributeKeyRef, existingVersionValueKey))
+                                         .expressionNames(Collections.singletonMap(attributeKeyRef, versionAttributeKey))
+                                         .expressionValues(Collections.singletonMap(existingVersionValueKey,
+                                                                                    existingVersionValue))
+                                         .build();
+
+        return Pair.of(newVersionValue, condition);
+    }
+
+    private int getExistingVersion(AttributeValue existingVersionValue) {
+        if (existingVersionValue.n() == null) {
+            throw new IllegalArgumentException("Version attribute appears to be the wrong type. N is required.");
+        }
+
+        return Integer.parseInt(existingVersionValue.n());
+    }
+
+    private AttributeValue incrementVersion(int version, DynamoDbExtensionContext.BeforeWrite context) {
+        Optional<Integer> versionIncrementByFromAnnotation = context.tableMetadata()
+                                                            .customMetadataObject(VersionAttribute.INCREMENT_BY_METADATA_KEY,
+                                                                                  Integer.class);
+        if (versionIncrementByFromAnnotation.isPresent()) {
+            return AttributeValue.fromN(Integer.toString(version + versionIncrementByFromAnnotation.get()));
+        }
+        return AttributeValue.fromN(Integer.toString(version + this.incrementBy));
+    }
+
     @NotThreadSafe
     public static final class Builder {
+        private int startAt = 0;
```

Comment:
Can we define the default value in the outer class instead of builder? IMO builder class should not have any logic